README.md
# README

This project contains a library which converts an ISO date/time string into a range.

## Usage

The module contains one function - `get_date_range` - which converts a given ISO
date/time string into a range and returns this range as a list. It is used as
follows:

```python
import isodaterange

range = isodaterange.get_date_range("2019-07")
# returns [datetime.datetime(2019, 7, 1, 0, 0, 0), datetime.datetime(2019, 7, 31, 23, 59, 59)]

range = isodaterange.get_date_range("2019-04-01/2019-06-23")
# returns [datetime.datetime(2019, 4, 1, 0, 0, 0), datetime.datetime(2019, 6, 23, 23, 59, 59)]

range = isodaterange.get_date_range("P1Y")
# returns [now, date one year from now]
```

## Testing

There is a `tests` module which contains unit tests. This can be run as follows:

```
python3 -m unittest isodaterange.tests
```


CHANGELOG.md
# CHANGELOG

## 1.0.1 [2020-05-07]

### Added

* Added `LICENSE.md`

### Changed

* Mirrored repository to Github
* Updated `setup.py` in order to upload to PyPI

## 1.0.0 [2020-04-28]

### Added

* (#1) - Added support for durations

## 0.1.1 [2020-01-16]

### Fixed

* Module actually installs correctly!

### Modified

* (#3) - Dates with no dates are now handled more efficiently

### Removed

* (#2) - Removed references to API project

## 0.1.0 [2019-12-04]

* First version


isodaterange/tests/__init__.py
################################################################################
# tests/dates.py
#
# Unit tests for MMP API
#
# TDBA 2019-11-15:
#   * First version
################################################################################
# CONFIGURATION
################################################################################
import datetime, unittest
import isodaterange
################################################################################
# TESTS
################################################################################
class DateTests(unittest.TestCase):
    def _test_date(self, date, range):
        test_range = isodaterange.get_date_range(date)
        self.assertEqual(test_range[0], range[0])
        self.assertEqual(test_range[1], range[1])
    def _test_duration(self, duration, range=None):
        now = datetime.datetime.now().replace(microsecond=0)
        new_range = {}
        for x in ["year", "month"]:
            if x in range:
                new_range[x] = getattr(now, x) + range[x]
        test_date = now.replace(**new_range)
        for x in ["day", "hour", "minute", "second"]:
            if x in range:
                attr = "{}s".format(x)
                test_date += datetime.timedelta(**{ attr: range[x] })
        test_range = isodaterange.get_date_range(duration)
        self.assertEqual(test_range[0], min([now, test_date]))
        self.assertEqual(test_range[1], max([now, test_date]))
    def test_range_year(self):
        self._test_date("2019",
            (datetime.datetime(2019, 1, 1, 0, 0, 0), datetime.datetime(2019, 12, 31, 23, 59, 59))
        )
    def test_range_year_month(self):
        self._test_date("2019-07",
            (datetime.datetime(2019, 7, 1, 0, 0, 0), datetime.datetime(2019, 7, 31, 23, 59, 59))
        )
    def test_range_date(self):
        self._test_date("2019-03-12",
            (datetime.datetime(2019, 3, 12, 0, 0, 0), datetime.datetime(2019, 3, 12, 23, 59, 59))
        )
    def test_range_datetime_hour(self):
        self._test_date("2016-04-18T12:00:00",
            (datetime.datetime(2016, 4, 18, 12, 0, 0), datetime.datetime(2016, 4, 18, 12, 0, 0))
        )
    def test_range_datetime_hr_min(self):
        self._test_date("2017-11-25T07:32:00",
            (datetime.datetime(2017, 11, 25, 7, 32, 0), datetime.datetime(2017, 11, 25, 7, 32, 0))
        )
    def test_range_datetime_hr_min_sec(self):
        self._test_date("2018-02-14T12:34:56",
            (datetime.datetime(2018, 2, 14, 12, 34, 56), datetime.datetime(2018, 2, 14, 12, 34, 56))
        )
    def test_range_year(self):
        self._test_date("2019/2020",
            (datetime.datetime(2019, 1, 1, 0, 0, 0), datetime.datetime(2020, 12, 31, 23, 59, 59))
        )
    def test_range_year_month(self):
        self._test_date("2019-07/2019-09",
            (datetime.datetime(2019, 7, 1, 0, 0, 0), datetime.datetime(2019, 9, 30, 23, 59, 59))
        )
    def test_range_date(self):
        self._test_date("2019-03-12/2019-04-25",
            (datetime.datetime(2019, 3, 12, 0, 0, 0), datetime.datetime(2019, 4, 25, 23, 59, 59))
        )
    def test_range_datetime_time(self):
        self._test_date("2016-04-18T12:00:00/2016-04-25T07:22:47",
            (datetime.datetime(2016, 4, 18, 12, 0, 0), datetime.datetime(2016, 4, 25, 7, 22, 47))
        )
    def test_range_datetime_december(self):
        self._test_date("2019-12",
            (datetime.datetime(2019, 12, 1, 0, 0, 0), datetime.datetime(2019, 12, 31, 23, 59, 59))
        )
    def test_range_to_infinity(self):
        self._test_date("2012-10-24/..",
            (datetime.datetime(2012, 10, 24, 0, 0, 0), None)
        )
    def test_range_from_infinity(self):
        self._test_date("../2012-10-24",
            (None, datetime.datetime(2012, 10, 24, 23, 59, 59))
        )
    def test_range_malformed(self):
        self.assertRaises(
            ValueError,
            lambda: isodaterange.get_date_range("fake-date")
        )
    def test_duration_year(self):
        self._test_duration("P1Y", { "year": 1 })
    def test_duration_month(self):
        self._test_duration("P2M", { "month": 2 })
    def test_duration_day(self):
        self._test_duration("P3D", { "day": 3 })
    def test_duration_week(self):
        self._test_duration("P1W", { "day": 7 })
    def test_duration_date_mixed(self):
        self._test_duration("P3Y6M4D", { "year": 3, "month": 6, "day": 4 })
    def test_duration_hour(self):
        self._test_duration("PT1H", { "hour": 1 })
    def test_duration_minute(self):
        self._test_duration("PT5M", { "minute": 5 })
    def test_duration_hour(self):
        self._test_duration("PT72S", { "second": 72 })
    def test_duration_time_mixed(self):
        self._test_duration("PT12H30M5S", { "hour": 12, "minute": 30, "second": 5})
    def test_duration_mixed(self):
        self._test_duration("P1Y2M10DT2H30M", { "year": 1, "month": 2, "day": 10, "hour": 2, "minute": 30})
    def test_duration_negative(self):
        self._test_duration("-P1Y", { "year": -1 })
    def test_duration_malformed(self):
        self.assertRaises(
            ValueError,
            lambda: isodaterange.get_date_range("Pfake-duration")
        )


isodaterange/__init__.py
################################################################################
# __init__.py
#
# Converts an ISO date string into a range
#
# TDBA 2019-11-25:
#   * First version
# TDBA 2019-12-20:
#   * Removed references to API (#2)
################################################################################
# CONFIGURATION
################################################################################
import datetime, isodate

# Get the date regular expressions and the default values
REGEXPS_DATE = isodate.isodates.build_date_regexps()
REGEXPS_TIME = isodate.isotime.build_time_regexps()
DEFAULTS = {
    "start": { "month": 1, "day": 1, "hour": 0, "minute": 0, "second": 0 },
    "end":   { "month": 12, "day": None, "hour": 23, "minute": 59, "second": 59 }
}
################################################################################
# FUNCTIONS
################################################################################
def get_date_range(d):
    """
    Converts a date string to a date range

    :param string d: Date to convert
    :returns: Start and end of date range
    :rtype: list
    """
    # If no date passed, return None
    if d is None:
        return None

    # If date begins with a "P" or a negative sign, it's a duration. Otherwise, it's a date
    if d.startswith(("P", "-")):
        return _parse_duration(d)
    else:
        return _parse_date(d)
def _parse_date(d):
    """
    Parses a date string

    :param string d: Date string to convert
    :returns: Start and end of date range
    :rtype: list
    """
    # Split the values on a slash
    dates = d.split("/")

    # Check we have a valid number of dates (i.e. 1 or 2)
    if len(dates) == 0 or len(dates) > 2:
        raise ValueError(
            "Multiple dates found in {}. Can only have one or two dates".format(d)
        )

    # If we have just one date, duplicate it
    if len(dates) == 1:
        dates.append(dates[0])

    # Convert the dates to datetime objects
    dt = [None, None]
    for idx, date_string in enumerate(dates):
        # If date consists of two dots (..), then it's an infinite bound.
        if date_string == "..":
            dt[idx] = None
            continue

        # Get the defaults
        defs = DEFAULTS["start"] if idx == 0 else DEFAULTS["end"]

        # Split the date string on "T" to obtain the date and time elements
        dt_parts = date_string.split("T")

        # Get the date part
        try:
            for r in REGEXPS_DATE:
                match = r.match(dt_parts[0])
                if match:
                    groups = match.groupdict()
                    year = int(groups["year"])
                    month = int(groups["month"]) if "month" in groups else defs["month"]
                    day = int(groups["day"]) if "day" in groups else defs["day"]
                    break

            # Fix day, if day is None
            if day is None:
                # Add 32 days to the first day of the month. This should bring us
                # into the next month
                temp_date = datetime.date(year, month, 1) + datetime.timedelta(days=32)

                # Get the first date of this new month and then subtract a single day
                temp_date = datetime.date(temp_date.year, temp_date.month, 1) - datetime.timedelta(days=1)

                # Get the day from the new temporary date
                day = temp_date.day

            # Assemble the date
            date = datetime.date(year, month, day)
        except Exception as e:
            raise ValueError("Cannot determine date from {}".format(date_string))

        # Get the time part, if it exists. Otherwise, use the defaults
        try:
            if len(dt_parts) == 2:
                for r in REGEXPS_TIME:
                    match = r.match(dt_parts[1])
                    if match:
                        groups = match.groupdict()
                        hour = int(groups["hour"]) if "hour" in groups else defs["hour"]
                        minute = int(groups["minute"]) if "minute" in groups else defs["minute"]
                        second = int(groups["second"]) if "second" in groups else defs["second"]
                        break
            else:
                hour = defs["hour"]
                minute = defs["minute"]
                second = defs["second"]

            # Aseemble the time
            time = datetime.time(hour, minute, second)
        except Exception as e:
            raise ValueError("Cannot determine time from {}".format(date_string))

        # Combine the date and time
        try:
            dt[idx] = datetime.datetime.combine(date, time)
        except Exception as e:
            raise ValueError("Cannot create date object from {}".format(date_string))

    # Return the dates
    return dt
def _parse_duration(d):
    """
    Parses a duration string

    :param string d: Duration string to convert
    :returns: Start and end of date range
    :rtype: list
    """
    try:
        # Get the duration
        duration = isodate.parse_duration(d)

        # Get the current date/time
        now = datetime.datetime.now().replace(microsecond=0)

        # Return tuple containing now and the date/time duration from now
        return sorted([now, now + duration])
    except Exception as e:
        raise ValueError("Cannot determine duration from {}".format(d))
